src/dataframerow/dataframerow.jl

```diff
@@ -86,6 +90,33 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
     isequal(cols1[1][r1], cols2[1][r2]) &&
         isequal_row(Base.tail(cols1), r1, Base.tail(cols2), r2)
 
+# table columns are passed as a tuple of vectors to ensure type specialization
+isequivalent_row(cols::Tuple{AbstractVector}, r1::Int, r2::Int) =
+    isequivalent(cols[1][r1], cols[1][r2])
+isequivalent_row(cols::Tuple{Vararg{AbstractVector}}, r1::Int, r2::Int) =
+    isequivalent(cols[1][r1], cols[1][r2]) && isequivalent_row(Base.tail(cols), r1, r2)
+
+isequivalent_row(cols1::Tuple{AbstractVector}, r1::Int, cols2::Tuple{AbstractVector}, r2::Int) =
+    isequivalent(cols1[1][r1], cols2[1][r2])
+isequivalent_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+            cols2::Tuple{Vararg{AbstractVector}}, r2::Int) =
+    isequivalent(cols1[1][r1], cols2[1][r2]) &&
+        isequivalent_row(Base.tail(cols1), r1, Base.tail(cols2), r2)
+
+function isequivalent(x::Any, y::Any)
+    contentmatches = x == y
+    # isa Missing should be redundant
+    # but it gives much more efficient code on Julia 0.6
+    if contentmatches isa Missing || ismissing(contentmatches)
+        if (x isa Missing || ismissing(x)) && (y isa Missing || ismissing(y))
```

Comment:
The main reason why I'm doing these checks is so that the behavior of joins and groupings with missings remain the same as it was before.

Reply:
I'm not even sure how to do this keeping the existing behaviour with missings? Since basically, we use `isequivalent` to check if 2 rows are equivalent so we can join on them, if one of the values in the row is missing we still need to somehow determine if the row itself is equivalent or not? I can leave this `Base.:(==)(r1::DataFrameRow, r2::DataFrameRow) = isequal(r1, r2)` without change, but I'm not sure how we can determine if 2 rows are equivalent if the result of a comparison check is a `missing` value.

Reply:
Yes, my proposal would break the existing behaviour. My understanding though is that other systems that use 3 valued logic would behave that way (rows with missing values are not equal). We may also want to thread the comparison operator through the API anyways?